syncer/syncer.go

```diff
@@ -0,0 +1,235 @@
+func (s *Syncer) blocklistFromPortal(portalURL string) ([]crypto.Hash, error) {
+	// convenience variables
+	logger := s.staticLogger
+
+	// build the request to fetch the blocklist from the portal
+	url := fmt.Sprintf("%s/skynet/blocklist", portalURL)
+	req, err := http.NewRequest(http.MethodGet, url, nil)
+	if err != nil {
+		return nil, errors.AddContext(err, "failed to build blocklist request")
+	}
+
+	// set headers and execute the request
+	req.Header.Set("User-Agent", "Sia-Agent")
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, errors.AddContext(err, "failed to fetch blocklist")
+	}
+	defer func() {
+		err = resp.Body.Close()
+		if err != nil {
+			logger.Errorf("failed to close the request body, err: %v", err)
+		}
+	}()
+
+	// decode the response
+	br := struct{ Blocklist []crypto.Hash }{}
+	err = json.NewDecoder(resp.Body).Decode(&br)
+	if err != nil {
+		return nil, errors.AddContext(err, "failed to decode skyd API response")
+	}
+
+	return br.Blocklist, nil
+}
+
+// buildLookupTable is a small helper function that fetches the blocklist from
+// the skyd API and turns it into a lookup table that can be used to see whether
+// a hash is part of the blocklist.
+func (s *Syncer) buildLookupTable() (lookupTable, error) {
+	blocklist, err := s.staticSkydAPI.Blocklist()
+	if err != nil {
+		return nil, err
+	}
+	lookup := make(map[crypto.Hash]struct{})
+	for _, hash := range blocklist {
+		lookup[hash] = struct{}{}
+	}
+	return lookup, nil
+}
+
+// syncDatabaseWithSkyd will diff the blocklist of the connected skyd with the
+// list of hashes in the database and add any missing hashes
+func (s *Syncer) syncDatabaseWithSkyd() error {
+	// convenience variables
+	logger := s.staticLogger
+
+	// build a lookup table of existing hashes
+	existing, err := s.buildLookupTable()
+	if err != nil {
+		return errors.AddContext(err, "could not build lookup table of existing hashes")
+	}
+
+	// fetch all hashes from our database
+	hashes, err := s.staticDB.Hashes()
+	if err != nil {
+		return errors.AddContext(err, "could not get hashes from database")
+	}
+
+	// filter out all hashes which are already blocked by skyd
+	var curr int
+	for _, hash := range hashes {
+		if _, exists := existing[hash.Hash]; !exists {
+			hashes[curr] = hash
+			curr++
+		}
+	}
+	hashes = hashes[:curr]
+
+	// use the blocker (has batching built in) to block the ones we are missing
+	added, _, err := s.staticBlocker.BlockHashes(hashes)
+	if err != nil {
+		return errors.AddContext(err, "could not block hashes")
+	}
+
+	logger.Infof("successfully synced hashes in database with skyd, %v hashes added", added)
+	return nil
+}
+
+// sync will loop through the preconfigured portals and pull their blocklist and
+// ensure all of the hashes on those blocklists are blocked in the local skyd
+// instance.
+func (s *Syncer) syncPortalWithSkyd(portalURL string, existing lookupTable) error {
```

Comment:
I don't think there is testing for this method.

Reply:
added